contracts/ecash/src/contract.rs

```diff
@@ -0,0 +1,430 @@
+            .ok_or(ContractError::Unauthorized)?;
+
+        if let Some(blacklist_proposal) = self
+            .blacklist_proposals
+            .may_load(ctx.deps.storage, &public_key)?
+        {
+            Ok(Response::new().add_attribute(
+                BLACKLIST_PROPOSAL_ID,
+                blacklist_proposal.proposal_id().to_string(),
+            ))
+        } else {
+            let msg = Self::create_blacklist_proposal(
+                public_key.to_string(),
+                ctx.env.contract.address.into_string(),
+                cfg.multisig_addr.into_string(),
+            )?;
+            Ok(Response::new().add_submessage(msg))
+        }
+    }
+
+    #[msg(exec)]
+    pub fn add_to_blacklist(
+        &self,
+        ctx: ExecCtx,
+        public_key: String,
+    ) -> Result<Response, ContractError> {
+        //Only by multisig contract, actually add public key to blacklist
+        self.admin
+            .assert_admin(ctx.deps.as_ref(), &ctx.info.sender)?;
+
+        self.blacklist.save(
+            ctx.deps.storage,
+            &public_key.clone(),
+            &BlacklistedAccount::new(public_key),
+        )?;
+        Ok(Response::new())
+    }
+    #[msg(reply)]
+    pub fn reply(&self, ctx: ReplyCtx, msg: Reply) -> Result<Response, ContractError> {
+        match msg.id {
+            BLACKLIST_PROPOSAL_REPLY_ID => self.handle_blacklist_proposal_reply(ctx, msg),
+            id => Err(ContractError::Std(cosmwasm_std::StdError::GenericErr {
+                msg: format!("Unknown reply Id {}", id),
+            })),
+        }
+    }
+
+    fn handle_blacklist_proposal_reply(
+        &self,
+        ctx: ReplyCtx,
+        msg: Reply,
+    ) -> Result<Response, ContractError> {
+        let reply = msg.result.into_result().map_err(StdError::generic_err)?;
+        let proposal_attribute = reply
+            .events
+            .iter()
+            .find(|event| event.ty == "wasm")
+            .ok_or(ContractError::ProposalError(
+                "Wasm event not found".to_string(),
+            ))?
+            .attributes
+            .iter()
+            .find(|attr| attr.key == BLACKLIST_PROPOSAL_ID)
+            .ok_or(ContractError::ProposalError(
+                "Proposal id not found".to_string(),
+            ))?;
+
+        let proposal_id = proposal_attribute.value.parse::<u64>().map_err(|_| {
+            ContractError::ProposalError(String::from("proposal id could not be parsed to u64"))
+        })?;
+
+        let cfg = self.config.load(ctx.deps.storage)?;
+        let msg = MultisigQueryMsg::Proposal { proposal_id };
+        let proposal_response: ProposalResponse = ctx.deps.querier.query(
+            &cosmwasm_std::QueryRequest::Wasm(cosmwasm_std::WasmQuery::Smart {
+                contract_addr: cfg.multisig_addr.to_string(),
+                msg: to_binary(&msg)?,
+            }),
+        )?;
+        let public_key = proposal_response.description;
```

Comment:
if we care about the pubkey, we don't have to query the multisig contract as we have the pubkey on hand when we created this submsg

Reply:
We do not, context does not propagate to replies
https://docs.cosmwasm.com/docs/smart-contracts/message/submessage/#propagation-of-context-between-contracts